src/main/host/syscall/handler/time.rs

```diff
@@ -96,4 +97,105 @@ impl SyscallHandler {
 
         Ok(0.into())
     }
+
+    #[log_syscall(/* clock_id */ linux_api::time::linux___kernel_clockid_t,
+        /* flags */ std::ffi::c_int, /* request */ *const std::ffi::c_void,
+        /* remain */ *const std::ffi::c_void)]
+    pub fn clock_nanosleep(
+        ctx: &mut SyscallContext,
+        clock_id: linux_api::time::linux___kernel_clockid_t,
+        flags: std::ffi::c_int,
+        request_ptr: ForeignPtr<linux_api::time::timespec>,
+        remain_ptr: ForeignPtr<linux_api::time::timespec>,
+    ) -> Result<std::ffi::c_int, SyscallError> {
+        let clock_id = ClockId::try_from(clock_id).map_err(|_| Errno::EINVAL)?;
+
+        // Check for clock_ids that specifically support nanosleep.
+        if [
+            ClockId::CLOCK_MONOTONIC,
+            ClockId::CLOCK_REALTIME,
+            ClockId::CLOCK_BOOTTIME,
+            ClockId::CLOCK_TAI,
+        ]
+        .contains(&clock_id)
+        {
+            // Supported and simulated in Shadow.
+            Self::nanosleep_helper(ctx, flags, request_ptr, remain_ptr)
+        } else if [ClockId::CLOCK_THREAD_CPUTIME_ID].contains(&clock_id) {
+            log::debug!("Invalid clock_id {clock_id:?}.",);
+            Err(Errno::EINVAL.into())
+        } else if [
+            ClockId::CLOCK_PROCESS_CPUTIME_ID, // Supported in Linux, not in Shadow.
+            ClockId::CLOCK_MONOTONIC_RAW,
+            ClockId::CLOCK_REALTIME_COARSE,
+            ClockId::CLOCK_MONOTONIC_COARSE,
+            ClockId::CLOCK_REALTIME_ALARM,
+            ClockId::CLOCK_BOOTTIME_ALARM,
+        ]
+        .contains(&clock_id)
+        {
+            log::debug!("Unsupported clock_id {clock_id:?}.",);
+            Err(Errno::ENOTSUP.into())
+        } else {
+            log::debug!("Unknown clock_id {clock_id:?}.");
+            Err(Errno::EINVAL.into())
+        }
+    }
+
+    fn nanosleep_helper(
+        ctx: &mut SyscallContext,
+        flags: std::ffi::c_int,
+        request_ptr: ForeignPtr<linux_api::time::timespec>,
+        remain_ptr: ForeignPtr<linux_api::time::timespec>,
+    ) -> Result<std::ffi::c_int, SyscallError> {
+        let request = ctx.objs.process.memory_borrow().read(request_ptr)?;
+        let request_time = SimulationTime::try_from(request).or(Err(Errno::EINVAL))?;
+        let flags = ClockNanosleepFlags::from_bits_truncate(flags);
```

Comment:
I don't think we want to silently truncate unknown flags here. e.g.

```
let Ok(flags) = ClockNanosleepFlags::from_bits(flags) else {
  debug!("Invalid flags {flags}");
  return Err(Errno::EINVAL.into())
};
```

Reply:
I think the Linux implementation only checks if `TIMER_ABSTIME` is set, it does not complain or return EINVAL if other bits are set too. Your suggestion would break from that behavior and return EINVAL if bits not defined in `ClockNanosleepFlags` are set.

I should have added a comment to explain why the truncate, so I'll do that.

Reply:
Yeah, looking at `kernel/time/posix-timers.c` I think you're right :)